### PR TITLE
#662  Exclude empty series on matrix page

### LINKED ISSUE #662: Hide empty series from the series list
https://www.stroyprombeton.ru/series/

### PR DESCRIPTION
**Pull request by duker33**
Closes #662

### CODE REVIEW

File: stroyprombeton/models.py [hunk 1]
```diff
@@ -81,13 +82,38 @@ def get_series(self) -> models.QuerySet:
 
 class SeriesQuerySet(models.QuerySet):
 
+    class SQLCount(Subquery):
+        template = "(SELECT count(*) FROM (%(subquery)s) _count)"
```
[the remark this answers is not in the document]
Comment (artemiy312) [reply stroyprombeton/models.py]: It would be nice move this class to outer scope

File: stroyprombeton/models.py [hunk 2]
```diff
@@ -81,13 +82,38 @@ def get_series(self) -> models.QuerySet:
...
+        #  `pages.models` module for example.
+        #  Don't forget to run over all "SQLCount" code entries.
+
+        # https://stackoverflow.com/questions/42543978/django-1-11-annotating-a-subquery-aggregate
+        class SQLCount(Subquery):
+            template = "(SELECT count(*) FROM (%(subquery)s) _count)"
+            output_field = IntegerField()
+
+        return (
+            self.active()
```
Comment (artemiy312) [inline stroyprombeton/models.py]: It seems it does the same, but less verbose and simpler
```suggestion
            self.active()
            .filter(options__product__page__is_active=True)
            .distinct()
```

File: stroyprombeton/models.py [hunk 3]
```diff
@@ -81,13 +82,38 @@ def get_series(self) -> models.QuerySet:
...
+
     def bind_fields(self):
         """Prefetch or select typical related fields to reduce sql queries count."""
         return self.select_related('page')
 
-    def active(self):
+    def active(self) -> 'SeriesQuerySet':
         return self.filter(page__is_active=True)
 
+    def non_empty(self) -> 'SeriesQuerySet':
```
Comment (artemiy312) [inline stroyprombeton/models.py]:
```suggestion
    def exclude_empty(self) -> 'SeriesQuerySet':
```

File: stroyprombeton/tests/tests_views.py [hunk 4]
```diff
@@ -947,6 +947,46 @@ def test_filter_options_by_tags(self):
...
+    def test_series_are_active(self):
+        """Every series presented in series list should be active."""
+        page = CustomPage.objects.get(slug='series')
+        response = self.client.get(page.url)
+        db_series = (
+            models.Series.objects.bind_fields()
+            .non_empty()
+            .order_by('name')
+        )
+        app_series = list(chain(*response.context['parted_series']))
```
Comment (artemiy312) [inline stroyprombeton/tests/tests_views.py]:
```suggestion
        app_series = list(chain.from_iterable(response.context['parted_series']))
```

File: stroyprombeton/tests/tests_views.py [hunk 5]
```diff
@@ -947,6 +947,46 @@ def test_filter_options_by_tags(self):
...
+            'html.parser'
+        )
+
+    def test_page(self):
+        self.assertEqual(200, self.get_page().status_code)
+
+    def test_series_are_active(self):
+        """Every series presented in series list should be active."""
+        page = CustomPage.objects.get(slug='series')
+        response = self.client.get(page.url)
```
Comment (artemiy312) [inline stroyprombeton/tests/tests_views.py]:
```suggestion
        response = self.get_page()
```

File: stroyprombeton/tests/tests_views.py [hunk 6]
```diff
@@ -947,6 +947,46 @@ def test_filter_options_by_tags(self):
...
+@tag('fast', 'catalog')
+class SeriesMatrix(BaseCatalogTestCase):
+    fixtures = ['dump.json']
+
+    def get_page_url(self):
+        page = CustomPage.objects.get(slug='series')
+        return page.url
+
+    def get_page(self, *args, **kwargs):
+        return self.client.get(self.get_page_url(*args, **kwargs))  # Ignore CPDBear
```
Comment (artemiy312) [inline stroyprombeton/tests/tests_views.py]: You should edit signature for `get_page_url` 
```python
    def get_page_url(self, *args, **kwargs):
```

File: stroyprombeton/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (duker33) [reply stroyprombeton/models.py]: @artemiy312 , created `@todo` for it at line ~99

File: stroyprombeton/tests/tests_views.py (hunk 6 above)
Comment (duker33) [reply stroyprombeton/tests/tests_views.py]: removed redundant `*args, **kwargs` passes instead

**Comment by artemiy312** [comment]
@duker33  It seems you forgot push fixes

**Comment by duker33** [comment]
@artemiy312 , i had problems with this PR and #669 code mixing. Now it's ok, you can review it

**Pull request merged**